src/input/mod.rs

```diff
@@ -0,0 +1,490 @@
+/// and `cluster_main()` as `run()` and `run_cluster()` will yield AdvanceTo
+/// and Emit for you. Likewise, they are only required when using a
+/// manual input configuration.
+///
+///
+/// See also: `inputs.yield_epochs()`
+///
+/// >>> def input_builder(worker_index, worker_count, resume_epoch):
+/// ...     for i in range(10):
+/// ...         yield AdvanceTo(i) # Advances the epoch to i
+/// ...         yield Emit(i) # Adds the input i at epoch i
+#[pyclass(module = "bytewax")]
+#[pyo3(text_signature = "(epoch)")]
+pub(crate) struct AdvanceTo {
+    #[pyo3(get)]
+    epoch: u64,
+}
+
+#[pymethods]
+impl AdvanceTo {
+    #[new]
+    fn new(epoch: u64) -> Self {
+        Self { epoch }
+    }
+
+    fn __richcmp__(&self, other: &Self, op: CompareOp) -> PyResult<bool> {
+        match op {
+            CompareOp::Lt => Ok(self.epoch < other.epoch),
+            CompareOp::Le => Ok(self.epoch <= other.epoch),
+            CompareOp::Eq => Ok(self.epoch == other.epoch),
+            CompareOp::Ne => Ok(self.epoch != other.epoch),
+            CompareOp::Gt => Ok(self.epoch > other.epoch),
+            CompareOp::Ge => Ok(self.epoch >= other.epoch),
+        }
+    }
+}
+
+/// Emit the supplied item into the dataflow at the current epoch
+///
+/// Emit is how we introduce input into a dataflow using a manual
+/// input configuration:
+///
+/// >>> def input_builder(worker_index, worker_count, resume_epoch):
+/// ...     for i in range(10):
+/// ...         yield AdvanceTo(i) # Advances the epoch to i
+/// ...         yield Emit(i) # Adds the input i at epoch i
+#[pyclass(module = "bytewax")]
+#[pyo3(text_signature = "(item)")]
+pub(crate) struct Emit {
+    #[pyo3(get)]
+    item: TdPyAny,
+}
+
+#[pymethods]
+impl Emit {
+    #[new]
+    fn new(item: Py<PyAny>) -> Self {
+        Self { item: item.into() }
+    }
+}
+
+/// Base class for an input config.
+///
+/// InputConfig defines how you will input data to your dataflow.
+///
+/// Use a specific subclass of InputConfig for the kind of input
+/// source you are plan to use. See the subclasses in this module.
+#[pyclass(module = "bytewax.inputs", subclass)]
+#[pyo3(text_signature = "()")]
+pub(crate) struct InputConfig;
+
+/// Use [Kafka](https://kafka.apache.org) as the input source.
+///
+///
+/// Args:
+///
+///     brokers: List of broker addresses.
+///
+///     group_id: Group id
+///
+///     topic: Topic to which consumer will subscribe
+///
+///     batch_size: Integer defining how messages will be grouped by epoch.
+///
+/// Returns:
+///
+///     Config object. Pass this as the `input_config` argument to
+///     your execution entry point.
+#[pyclass(module = "bytewax.inputs", extends = InputConfig)]
+#[pyo3(text_signature = "(brokers, group_id, topics)")]
+pub(crate) struct KafkaInputConfig {
+    #[pyo3(get)]
+    pub brokers: String,
+    #[pyo3(get)]
+    pub group_id: String,
+    #[pyo3(get)]
+    pub topics: String,
+    #[pyo3(get)]
+    pub batch_size: u64,
```

Comment:
`epoch_size`?

Reply:
I don't know. I'm undecided on our semantics around epochs I think. I started with `messages_per_epoch` but it felt clunky. I'll revisit it more